prow/plugins/setmilestone/milestone.go

```diff
@@ -0,0 +1,132 @@
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "milestone"
+
+var (
+	milestoneRegex   = regexp.MustCompile(`(?m)^/milestone\s+(.+)$`)
+	mustBeSigLead    = "You must be a member of the [kubernetes/kubernetes-milestone-maintainers](https://github.com/orgs/kubernetes/teams/kubernetes-milestone-maintainers/members) github team to set the milestone."
+	invalidMilestone = "The provided milestone is not valid for this repository.  Here are the available milestones:\n %s"
+)
+
+type githubClient interface {
+	CreateComment(owner, repo string, number int, comment string) error
+	SetMilestone(org, repo string, issueNum, milestoneNum int) error
+	ListTeamMembers(id int) ([]github.TeamMember, error)
+	ListMilestones(org, repo string) ([]github.Milestone, error)
+}
+
+func init() {
+	plugins.RegisterGenericCommentHandler(pluginName, handleGenericComment, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The milestone plugin allows members of the milestone maintainers Github team to set the milestone to an issue or pull request pull request.",
+		Config: map[string]string{
+			"": fmt.Sprintf("The milestone maintainers team is the Github team with ID: %d.", config.MilestoneStatus.MaintainersID),
+		},
+	}
+	pluginHelp.AddCommand(pluginhelp.Command{
+		Usage:       "/milestone <milestone number>",
+		Description: "Updates the milestone for an issue or PR",
+		Featured:    false,
+		WhoCanUse:   "Members of the milestone maintainers Github team can use the '/milestone' command. This team is specified in the config by providing the Github team's ID.",
+		Examples:    []string{"/milestone 1", "/milestone 38"},
+	})
+	return pluginHelp, nil
+}
+
+func handleGenericComment(pc plugins.PluginClient, e github.GenericCommentEvent) error {
+	return handle(pc.GitHubClient, pc.Logger, &e, pc.PluginConfig.MilestoneStatus.MaintainersID)
+}
+
+func buildMilestoneMap(milestones []github.Milestone) map[string]int {
+	m := make(map[string]int)
+	for _, ms := range milestones {
+		m[ms.Title] = ms.Number
+	}
+	return m
+}
+func handle(gc githubClient, log *logrus.Entry, e *github.GenericCommentEvent, maintainersID int) error {
+	if e.Action != github.GenericCommentActionCreated {
+		return nil
+	}
+
+	milestoneMatch := milestoneRegex.FindStringSubmatch(e.Body)
+	if milestoneMatch == nil {
+		return nil
+	}
+
+	org := e.Repo.Owner.Login
+	repo := e.Repo.Name
+
+	milestoneMaintainers, err := gc.ListTeamMembers(maintainersID)
+	if err != nil {
+		return err
+	}
+	found := false
+	for _, person := range milestoneMaintainers {
+		login := strings.ToLower(e.User.Login)
+		if strings.ToLower(person.Login) == login {
+			found = true
+			break
+		}
+	}
+	if !found {
+		// not in the milestone maintainers team
+		return gc.CreateComment(org, repo, e.Number, mustBeSigLead)
+	}
+
+	milestones, err := gc.ListMilestones(org, repo)
```

Comment:
should we list these every time instead of just catching the appropriate API error and posting a better message? this seems like an unnecessary use of tokens.
cc @cjwagner

Reply:
The milestones need to be listed to get the milestone number for the call to `SetMilestone`.

Reply:
Ah, I thought we were still requiring the milestone number per the usage. in that case we should probably update / clarify the command usage?

Reply:
@grodrigues3 ^